### ISSUE TITLE
subcommand callback executes twice

### ISSUE THREAD

**Issue by phlptp** [henryiii/CLI11#670]
The following code executes the g1c1 callback twice

```
#include <CLI/CLI.hpp>
#include <iostream>

int main(int argc, char* argv[]) {
  auto app = CLI::App{"An app."};
  auto group1 = app.add_option_group("FirstGroup", "First options group.");
  auto group2 = app.add_option_group("SecondGroup", "Second options group.");

  auto g1c1 = group1->add_subcommand("g1c1", "First group first sub-command")
                  ->callback([]() {
                    std::cout << "First group first sub-command executed.\n";
                  });

  auto g2c1 = group2->add_subcommand("g2c1", "Second group first sub-command");

  CLI11_PARSE(app, argc, argv);
  return 0;
}
```

This bug was reported by @bimalgaudel on [gitter](https://gitter.im/CLI11gitter/Lobby?at=618dd0d0fb8ca0572b1862e8)

**Comment by phlptp** [comment]
One question for the fix is the order in which callbacks execute.

I put a fix in #666, that that guarantees an option group callback would execute before any of its subscommand callbacks, that seems to be the most appropriate, but it also means that in the case of multiple subcommands all embedded in option groups they don't execute in order they are received but in the order of option group definitions.  

Either way has some potential issues.  

A final, somewhat more complicated to implement way, might be to order the options groups based on subcommand callbacks.

In most cases this will not be important,  I haven't seen many cases with mixed option group callbacks and subcommand callbacks which is where this discussion becomes relevant.

**Comment by bimalgaudel** [comment]
> One question for the fix is the order in which callbacks execute.
> 
> I put a fix in #666, that that guarantees an option group callback would execute before any of its subscommand callbacks, that seems to be the most appropriate, but it also means that in the case of multiple subcommands all embedded in option groups they don't execute in order they are received but in the order of option group definitions.
> 
> Either way has some potential issues.
> 
> A final, somewhat more complicated to implement way, might be to order the options groups based on subcommand callbacks.
> 
> In most cases this will not be important, I haven't seen many cases with mixed option group callbacks and subcommand callbacks which is where this discussion becomes relevant.

Indeed, the order of sub-commands is relevant in many cases when we have multiple sub-commands. For example `myprog.o (generate | read) [OPTS for group1] (report | format) [OPTS for group2]`.

Some considerations:
  -- Maybe it's a good idea to restrict the order in which options related to a group of sub-commands can appear.
      Eg. `generate --repopt  report --genopt` is ill-formed.
  -- Probably, it's already implemented this way, but first validating the command-line as a whole before calling any call-backs is a must. For instance, one doesn't want to execute an expensive callback on `generate` sub-command only to exit with a failure because the options to `report` sub-command were ill-formed.

**Comment by phlptp** [comment]
assuming you are using normal callbacks the error checking is completed by the time the callbacks are executed.  

I suppose having them in option groups gives you a way to control the order of subcommand callbacks.  since subcommand callbacks in option groups would now get executed in order of option group definition in the code vs order of user specification on the command line.    Within a single group it would be order of user specification.  

for example 
```
myprog generate --opt1 report --opt1  
```
would execute callbacks in the same order as 
```
myprog report --opt1  generate --opt1 
```
if they are defined in option groups where as the order would be different if they were all in main app
